Rubberduck.RibbonDispatcher/RibbonFactory.cs

```diff
@@ -0,0 +1,175 @@
+    /// values" usages are (believed to be) the only means of implementing functionality equivalent
+    /// to "overrides" in a COM-compatible way.
+    /// </remarks>
+    [Serializable]
+    [CLSCompliant(true)]
+    [ComDefaultInterface(typeof(IRibbonFactory))]
+    public class RibbonFactory : IRibbonFactory {
+        internal RibbonFactory(IRibbonUI ribbonUI, ResourceManager resourceManager) {
+            _ribbonUI        = ribbonUI;
+            _controls        = new Dictionary<string, IRibbonCommon>();
+            _buttons         = new Dictionary<string, IActionItem>();
+            _toggles         = new Dictionary<string, IToggleItem>();
+            _dropDowns       = new Dictionary<string, IDropDownItem>();
+            _imageables      = new Dictionary<string, IImageableItem>();
+            _resourceManager = resourceManager;
+        }
+
+        private readonly IRibbonUI                            _ribbonUI;
+        private readonly ResourceManager                      _resourceManager;
+        private readonly IDictionary<string, IRibbonCommon>   _controls;
+        private readonly IDictionary<string, IActionItem>     _buttons;
+        private readonly IDictionary<string, IDropDownItem>   _dropDowns;
+        private readonly IDictionary<string, IImageableItem>  _imageables;
+        private readonly IDictionary<string, IToggleItem>     _toggles;
+
+        /// <summary>Returns a readonly collection of all Ribbon Controls in this Ribbon ViewModel.</summary>
+        public IReadOnlyDictionary<string, IRibbonCommon>  Controls   => new ReadOnlyDictionary<string, IRibbonCommon>(_controls);
+        /// <summary>Returns a readonly collection of all Ribbon (Action) Buttons in this Ribbon ViewModel.</summary>
+        public IReadOnlyDictionary<string, IActionItem>    Buttons    => new ReadOnlyDictionary<string, IActionItem>(_buttons);
+        /// <summary>Returns a readonly collection of all Ribbon DropDowns in this Ribbon ViewModel.</summary>
+        public IReadOnlyDictionary<string, IDropDownItem>  DropDowns  => new ReadOnlyDictionary<string, IDropDownItem>(_dropDowns);
+        /// <summary>Returns a readonly collection of all Ribbon Imageable Controls in this Ribbon ViewModel.</summary>
+        public IReadOnlyDictionary<string, IImageableItem> Imageables => new ReadOnlyDictionary<string, IImageableItem>(_imageables);
+        /// <summary>Returns a readonly collection of all Ribbon Toggle Buttons in this Ribbon ViewModel.</summary>
+        public IReadOnlyDictionary<string, IToggleItem>    Toggles    => new ReadOnlyDictionary<string, IToggleItem>(_toggles);
+
+        private void PropertyChanged(object sender, IControlChangedEventArgs e) => _ribbonUI.InvalidateControl(e.ControlId);
+
+        private T Add<T>(T ctrl) where T:RibbonCommon {
+            _controls.Add(ctrl.Id, ctrl);
+            var button    = ctrl as IActionItem;    if (button    != null) _buttons   .Add(ctrl.Id, button);
```

Comment:
these assignments can be simplified with the  C# 7 features we can use since #3346 into:

    if (ctrl is IActionItem button) { _buttons.add(ctrl.Id, button); }

Reply:
Then again ... I'm not entirely sold on the idea that we should expose access to RibbonControls by Type...

Reply:
New instruction belongs on a new line; the whitespace looks weird. Are you constantly fighting the IDE to get that layout? If I overwrite the final semicolon here, my VS would remove all the extraneous whitespace. Also we avoid implicit scopes; all scopes need their curly braces.

    var button    = ctrl as IActionItem;    if (button    != null) _buttons   .Add(ctrl.Id, button);

Should be:

    var button = ctrl as IActionItem;
    if (button != null)
    {
        _buttons.Add(ctrl.Id, button);
    }

Reply:
COM Events are only exposed on classes, not on interfaces, since COM doesn't do delegates.

Reply:
Not sure what that has to do with whitespace and implicit scopes... am I missing something?

Reply:
@retailcoder : The COM response was to @Vogel612 's second comment. With C# 7 okay I will update my VS. No implicit scopes on a single line seems excessive, but I will comply.